Services/Diagnostics/DiagnosticsLogger.cs

```diff
@@ -0,0 +1,160 @@
+                                                null);
+
+            return await httpClient.PostAsync(this.PrepareRequest(this.servicesConfig.DiagnosticsEndpointUrl, jsonStruct));
+        }
+
+        public async Task<IHttpResponse> LogServiceHeartbeatAsync()
+        {
+            JsonStruct jsonStruct = new JsonStruct(Guid.NewGuid().ToString(),
+                                                SERVICE_HEARTBEAT_EVENT,
+                                                DEPLOYMENT_ID,
+                                                SOLUTION_TYPE,
+                                                null);
+
+            return await httpClient.PostAsync(this.PrepareRequest(this.servicesConfig.DiagnosticsEndpointUrl, jsonStruct));
+        }
+
+        public async Task<IHttpResponse> LogServiceErrorAsync(
+            string message,
+            [CallerMemberName] string callerName = "",
+            [CallerFilePath] string filePath = "",
+            [CallerLineNumber] int lineNumber = 0)
+        {
+            JsonStruct jsonStruct = this.ConvertToJson(message, null, null, callerName, filePath, lineNumber);
+            return await httpClient.PostAsync(this.PrepareRequest(this.servicesConfig.DiagnosticsEndpointUrl, jsonStruct));
+        }
+
+        public async Task<IHttpResponse> LogServiceErrorAsync(
+            string message,
+            Exception e,
+            [CallerMemberName] string callerName = "",
+            [CallerFilePath] string filePath = "",
+            [CallerLineNumber] int lineNumber = 0)
+        {
+            JsonStruct jsonStruct = this.ConvertToJson(message, e, null, callerName, filePath, lineNumber);
+            return await httpClient.PostAsync(this.PrepareRequest(this.servicesConfig.DiagnosticsEndpointUrl, jsonStruct));
+        }
+
+        public async Task<IHttpResponse> LogServiceErrorAsync(
+            string message,
+            object data,
+            [CallerMemberName] string callerName = "",
+            [CallerFilePath] string filePath = "",
+            [CallerLineNumber] int lineNumber = 0)
+        {
+            JsonStruct jsonStruct = this.ConvertToJson(message, null, data, callerName, filePath, lineNumber);
+            return await httpClient.PostAsync(this.PrepareRequest(this.servicesConfig.DiagnosticsEndpointUrl, jsonStruct));
+        }
+
+        private JsonStruct ConvertToJson(string message,
+            Exception e,
+            object data,
+            [CallerMemberName] string callerName = "",
+            [CallerFilePath] string filePath = "",
+            [CallerLineNumber] int lineNumber = 0)
+        {
+            Dictionary<string, object> eventProps = new Dictionary<string, object>();
+            eventProps.Add("message", message + $"(CallerMemberName = '{callerName}', CallerFilePath = '{filePath}', lineNumber = '{lineNumber}')");
+            if (e != null)
+            {
+                eventProps.Add("Exception", e);
```

Comment:
do we need to send full exception to Diagnostics or just Exception message is fine? For Diagnostics - exception message alone should be enough alongwith filePath and lineNumber
@dluc/ @troyhopwood  - Can you confirm?